src/experiment_prototype/experiment_prototype.py

```diff
@@ -951,3 +954,223 @@ def self_check(self):
+        """
+        Calculates the closest actual center frequency based on the USRP
+        device clock and the desired center frequency.
+
+        :param  chosen_freq: the center frequency desired
+        :type   chosen_freq: float
+
+        :return actual_center_freq: valid center frequency for USRP device
+        :rtype  actual_center_freq: float
+        """
+        clock_multiples = self.options.usrp_master_clock_rate / 2**32
+        clock_divider = np.ceil(chosen_freq * 1e3 / clock_multiples)
+        return (clock_divider * clock_multiples) / 1e3
+
+    def set_center_frequencies(self):
+        """
+        Determines and sets a tx and rx center frequency for any slices the user did not set manually
+
+        :raises ExperimentException:    if while loop continues for too long (10000 iterations)
+        """
+
+        # Initialize parameters
+        slice_freq, slice_ctr_freq = dict(), dict()
+        accounted_for, strict_slices = set(), set()
+        ref_set = frozenset([x for x in self.slice_ids])
+
+        tx_null_band = 50  # kHz
+        lowest_freq_slice, counter, center_freq = 0, 0, 0  # initial value unused
+        tune_bandwidth = (self.tx_bandwidth - transition_bandwidth * 2) / 1000  # kHz
+
+        down_shift = False
+        down_shifting_slice = None
+
+        # Create list of slices that need center frequencies and the freq used by the slice
+        for slice_id in self.slice_ids:
+            if (
+                self.slice_dict[slice_id].txctrfreq is None
+                and self.slice_dict[slice_id].rxctrfreq is not None
+            ):
+                self.slice_dict[slice_id].txctrfreq = self.slice_dict[
+                    slice_id
+                ].rxctrfreq
+            elif (
+                self.slice_dict[slice_id].txctrfreq is not None
+                and self.slice_dict[slice_id].rxctrfreq is None
+            ):
+                self.slice_dict[slice_id].rxctrfreq = self.slice_dict[
+                    slice_id
+                ].txctrfreq
+            elif (
+                self.slice_dict[slice_id].txctrfreq is None
+                and self.slice_dict[slice_id].rxctrfreq is None
+            ):
+                if self.slice_dict[slice_id].freq is None:
+                    slice_freq[slice_id] = self.slice_dict[slice_id].cfs_range
+                else:
+                    slice_freq[slice_id] = [self.slice_dict[slice_id].freq]
+
+        # If no frequencies need to be set
+        if len(slice_freq) == 0:
+            return
+
+        # determine which slices must have the same center frequencies
+        for slice_id in self.slice_ids:
+            interfacing_dict = self.get_slice_interfacing(slice_id)
+            interfacing_dict[slice_id] = (
+                "CONCURRENT"  # add current slice to interfacing dict
+            )
+            keys = list(interfacing_dict.keys())
+            keys.append(slice_id)
+            strict_slices.add(
```

Comment:
`strict_slices` is a set so if the fronzenset `{0, 1}` is already in `strict_slices` then it will not be added again. So yes overlaps occur but only each unique combination is kept. Also because they are sets, `{0, 1}` is considered the same as `{1, 0}`.